bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java

```diff
@@ -0,0 +1,187 @@
+            return socket;
+        }
+
+        private ServerInfo tryParseServerDiscovery(DatagramPacket receivePacket) throws ParseException {
+            String host = receivePacket.getAddress().getHostAddress();
+            String reply = new String(receivePacket.getData()).replaceAll("[\\p{C}]", NON_PRINTABLE_CHARTS_REPLACEMENT)
+                    .replaceAll("[^\\x00-\\x7F]", NON_PRINTABLE_CHARTS_REPLACEMENT);
+            if (!reply.startsWith(DISCOVERY_RESPONSE_PREFIX))
+                throw new ParseException("Bad discovery response prefix", 0);
+            String[] parts = Arrays
+                    .stream(reply.replace(DISCOVERY_RESPONSE_PREFIX, "").split(NON_PRINTABLE_CHARTS_REPLACEMENT))
+                    .filter((String e) -> e.length() != 0).toArray(String[]::new);
+            String name = parts[0];
+            int tcpPort = Integer.parseInt(parts[1]);
+            int udpPort = Integer.parseInt(parts[3]);
+            String macAddress = parts[2];
+            String publicIp = parts[4];
+            return new ServerInfo(host, tcpPort, udpPort, name, macAddress, publicIp);
+        }
+
+        /**
+         * Send broadcast packets with service request string until a response
+         * is received. Return the response as String (even though it should
+         * contain an internet address).
+         *
+         * @return String received from server. Should be server IP address.
+         *         Returns empty string if failed to get valid reply.
+         */
+        public void call(Consumer<ServerInfo> listener) {
+            byte[] receiveBuffer = new byte[MAX_PACKET_SIZE];
+            DatagramPacket receivePacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
+
+            DatagramSocket socket = null;
+            try {
+                socket = createSocket();
+            } catch (SocketException e) {
+                logger.error("Error creating socket: {}", e.getMessage());
+                return;
+            }
+            byte[] packetData = DISCOVERY_REQUEST.getBytes();
+            try {
+                InetAddress broadcastAddress = InetAddress.getByName("255.255.255.255");
+                int servicePort = DISCOVERY_PORT;
+                DatagramPacket packet = new DatagramPacket(packetData, packetData.length, broadcastAddress,
+                        servicePort);
+                socket.send(packet);
+                logger.debug("Sent packet to {}:{}", broadcastAddress.getHostAddress(), servicePort);
+                for (int i = 0; i < 20; i++) {
+                    socket.receive(receivePacket);
+                    String host = receivePacket.getAddress().getHostAddress();
+                    logger.debug("Received reply from {}", host);
+                    try {
+                        ServerInfo serverInfo = tryParseServerDiscovery(receivePacket);
+                        listener.accept(serverInfo);
+                    } catch (Exception ex) {
```

Comment:
You should specify the concrete exception you want to catch, since catching `Exception` catches also all runtime exceptions.